### PR TITLE
Dynamic project routes + APIv2 Support

### PR DESCRIPTION
**Pull request by venashial**
Support for the v2 API should be complete except for a few more things that need to be implemented on labrinth. They are:
- Allow uploading ZIPs

(Sorry for making this a separate PR than #250, not sure how else to do it)
(Also Geometrically already locally implemented API based categories iirc)

### CODE REVIEW

**Comment by venashial** [comment]
Here's a deploy: https://knossos-venashial-7zawv7rtn-venashial.vercel.app/

**Comment by venashial** [comment]
Updated deploy: https://knossos-venashial-7ulcovjsw-venashial.vercel.app/

**Comment by venashial** [comment]
Updated deploy: https://knossos-venashial-qsky5l87i-venashial.vercel.app/
(Fixed user pages)

File: components/pages/project/version/_version/index.vue [hunk 1]
```diff
@@ -297,7 +297,11 @@ export default {
 
       await this.$axios.delete(`version/${this.version.id}`, this.$auth.headers)
 
-      await this.$router.replace(`/mod/${this.mod.id}`)
+      await this.$router.replace(
+        `/${this.project.project_type}/${
+          this.project.slug ? this.project.slug : this.project.id
+        }`
+      )
```
Comment (wafflecoffee) [inline components/pages/project/version/_version/index.vue:304]: https://knossos-venashial-qsky5l87i-venashial.vercel.app/mod/fred/version/nTPy9wmt
^ This works, whereas
https://knossos-venashial-qsky5l87i-venashial.vercel.app/mod/fred/version/v1.fred
gives `500: INTERNAL_SERVER_ERROR Code: FUNCTION_INVOCATION_FAILED`.

**Review by wafflecoffee — CHANGES_REQUESTED** [review]
I reviewed Geo's/Red's PR before it was done, so I might as well do the same with yours :P

FYI, you should try pulling 46bc04f69f40b4ec54784aede4f5937e38bfbec6 into this branch to resolve conflicts, preferring my "phrasings" (with the added caveat that mods should be -> projects).

**Comment by venashial** [comment]
Updated deploy: https://knossos-venashial-b4kwdf0n4-venashial.vercel.app/

File: components/pages/project/version/_version/index.vue (hunk 1 above)
Comment (venashial) [reply components/pages/project/version/_version/index.vue:309]: Not sure what caused the error, but it appears to be fixed in the latest deploy

File: components/pages/project.vue [hunk 2]
```diff
@@ -415,29 +433,16 @@ export default {
...
-          data.$auth.headers
-        )
-      ).data
-
-      users.forEach((it) => {
-        const index = members.findIndex((x) => x.user_id === it.id)
-        members[index].avatar_url = it.avatar_url
-        members[index].name = it.username
-      })
-
```
[the remark this answers is not in the document]
Comment (venashial) [reply components/pages/project.vue:430]: That had to do with APIv2 changes resulting in not needing to get user data in a second request

File: components/pages/project/edit.vue [hunk 3]
```diff
@@ -32,27 +29,31 @@
...
       <h3>Categories</h3>
       <label>
         <span>
-          Select up to 3 categories. They will help to find your mod
+          Select up to 3 categories. They will help to find your project
         </span>
         <multiselect
           id="categories"
-          v-model="mod.categories"
+          v-model="project.categories"
```
[the remark this answers is not in the document]
Comment (venashial) [reply components/pages/project/edit.vue]: With the latest commits, it still has this issue, but it is rooted in the `tags` store. I'll need to sort the `$tags.categories` into `$tags[projectType].categories` or something like that. I'm waiting for the API to add modpack categories

File: static/projectRoutes.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  {
+    "type": "modpack",
+    "search": {
+      "name": "Modpacks",
+      "path": "modpacks",
+      "filters": [
+        "categories",
+        "mod_loader",
+        "eviroment",
```
[the remark this answers is not in the document]
Comment (venashial) [reply static/projectRoutes.json]: Resolved, though I still haven't made environment filters (or any filters) depend on the project type.

File: static/projectRoutes.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
+[
+  {
+    "type": "mod",
+    "search": {
+      "name": "Mods",
+      "path": "mods",
+      "filters": [
+        "categories",
+        "mod_loader",
+        "eviroment",
```
[the remark this answers is not in the document]
Comment (venashial) [reply static/projectRoutes.json]: Resolved

File: components/pages/project.vue [hunk 6]
```diff
@@ -478,19 +492,19 @@ export default {
       elem.href = url
       elem.click()
     },
-    async followMod() {
+    async followProject() {
       await this.$axios.post(
-        `mod/${this.mod.id}/follow`,
+        `${this.project.id}/follow`,
```
[the remark this answers is not in the document]
Comment (venashial) [reply components/pages/project.vue]: Fixed

File: components/pages/project.vue [hunk 7]
```diff
@@ -449,10 +454,19 @@ export default {
     } catch {
       data.error({
         statusCode: 404,
-        message: 'Mod not found',
+        message: 'Project not found',
```
[the remark this answers is not in the document]
Comment (venashial) [reply components/pages/project.vue:456]: In `asyncData` there isn't any way to access `this` which means no props, data, or route. I can't think of another way of implementing it

File: pages/create/project.vue [hunk 8]
```diff
@@ -756,24 +794,24 @@ export default {
...
-    'description  description description' auto
-    'versions     versions    versions' auto
-    'extra-links  extra-links extra-links' auto
-    'license      license     license' auto
-    'donations    donations   donations' auto
-    'footer       footer      footer' auto
+    'header       header       header' auto
+    'advert       advert       advert' auto
+    'essentials   essentials   essentials' auto
+    'project-icon project-icon  project-icon' auto
```
[the remark this answers is not in the document]
Comment (venashial) [reply pages/create/project.vue]: Fixed. I don't think this is something Prettier can do

**Comment by venashial** [comment]
Updated deploy: https://knossos-venashial-zyvcdyjkv-venashial.vercel.app/

File: components/pages/project.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (impossiblereality) [reply components/pages/project.vue:456]: maybe use `fetch()`?

**Comment by venashial** [comment]
Updated deploy: https://knossos-venashial-i9sf6d85b-venashial.vercel.app/

I added `datapacks` as a project route to better test project types.

File: components/pages/project.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (venashial) [reply components/pages/project.vue:456]: I attempted using `fetch()` but it doesn't seem like it is made for this type of thing. `fetch()` is better for getting a nonessential part of the page that scales, not something fundamental. The page still tried loading before fetch had finished, and it just errored. I could have been doing it wrong though.

**Comment by venashial** [comment]
TODO:
- [ ] Sort tags (etc. categories, mod loaders) by project type
- [ ] Add project type to the `ProjectCard` component on pages like dashboard and user

**Pull request closed without merging**